util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -74,195 +93,221 @@ var (
 	return &Worker{}
 }
 
 // Start starts the worker.
 func (w *Worker) Start() {
-	w.stopCh = make(chan struct{})
-	// TODO(#1631): make controller and worker communicate using k8s API (with CustomResources)
-	// instead of http.
-	w.listenToController()
+	w.initialize()
+	w.populatePodName()
+	w.registerForCustomResourceEvents(w.handleCustomResource)
 }
 
-func (w *Worker) listenToController() {
-	handlers := handlersMap{
-		"/startTCPServer":  w.StartTCPServer,
-		"/startTCPClient":  w.StartTCPClient,
-		"/startUDPServer":  w.StartUDPServer,
-		"/startUDPClient":  w.StartUDPClient,
-		"/startHTTPServer": w.StartHTTPServer,
-		"/startHTTPClient": w.StartHTTPClient,
-		"/metrics":         w.Metrics,
+func (w *Worker) initialize() {
+	w.gvr, _ = meta.UnsafeGuessKindToResource(gvk)
+	var err error
+	w.k8sClient, err = getDynamicClient()
+	if err != nil {
+		klog.Fatalf("Error getting dynamic client:%s", err)
 	}
-	w.startListening(workerListenPort, handlers)
+	w.resourceInterface = w.k8sClient.Resource(w.gvr).Namespace(namespace)
 }
 
-func (w *Worker) startListening(port string, handlers handlersMap) {
-	for urlPath, handler := range handlers {
-		http.HandleFunc(urlPath, handler)
-	}
-	if err := http.ListenAndServe(":"+port, nil); err != nil {
-		klog.Fatalf("Failed to start http server on port %v: %v", port, err)
+func (w *Worker) populatePodName() {
+	var isPodNameAvailable bool
+	w.podName, isPodNameAvailable = os.LookupEnv("POD_NAME")
+	if !isPodNameAvailable {
+		err := errors.New("pod name not set as environment variable")
+		klog.Fatal(err)
 	}
+	klog.Info("Pod Name set:", w.podName)
 }
 
-// Metrics returns the metrics collected.
-func (w *Worker) Metrics(rw http.ResponseWriter, request *http.Request) {
-	select {
-	case <-w.stopCh:
-		if w.err != nil {
-			message := fmt.Sprintf("metrics collection failed: %v", w.err)
-			w.sendResponse(rw, http.StatusInternalServerError, Status{Message: message})
-			return
-		}
-		reply := MetricResponse{
-			Result:      w.parsedResult,
-			WorkerDelay: w.workerDelay,
-		}
-		w.sendResponse(rw, http.StatusOK, reply)
-	default:
-		w.sendResponse(rw, http.StatusInternalServerError, Status{Message: "metric collection in progress"})
+func (w *Worker) registerForCustomResourceEvents(handlerFunc func(interface{})) {
+	informer, err := getInformer(w.getCustomResourceLabelSelector(), namespace, w.k8sClient, w.gvr)
+	if err != nil {
+		klog.Fatalf("Error getting informer:%s", err)
 	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			handlerFunc(obj)
+		},
+	})
+	w.stopCh = make(chan struct{})
+	informer.Run(w.stopCh)
 }
 
-func (w *Worker) sendResponse(rw http.ResponseWriter, statusCode int, response interface{}) {
-	rw.Header().Set("Content-Type", "application/json")
-	marshalledResponse, err := json.Marshal(response)
+func (w *Worker) getCustomResourceLabelSelector() string {
+	return fmt.Sprintf("%s in (clientPodName,serverPodName)", w.podName)
```

Comment:
Is it really what you want?

Reply:
Yes. 
The CR will have <podname>: clientPodName  <podname>: serverPodName. Had to put it this way since OR condition doesn't seem to be supported for labelselector(if it was we would have put it the other way ie clientPodName:<podName>  serverPodName:<podName> and could have put clientPOdNAme = <podName> OR serverPodName=<podNAme>). I suppose you were pointing to the same .